### DISCUSSION TITLE
Begin Equals on records with check for equality

### DISCUSSION THREAD

**Discussion by Xenoage** [dotnet/csharplang#4411] [Language Ideas]
Currently, the auto-generated `Equals` method for records look like this:

```c#
public virtual bool Equals(MyRecord? other) {
   return (object)other != null
     && EqualityContract == other!.EqualityContract
     && EqualityComparer<T>.Default.Equals(FirstProp, other!.FirstProp)
     && ...
}
```

Since `record`s will be used a lot for immutable data structures, which encourages to reuse and share objects, it would be very helpful to check for reference equality first. In this case, the `Equals` method would be speeded up enormously, especially for large object trees. Like this:

```c#
public virtual bool Equals(MyRecord? other) {
   if (object.ReferenceEquals(this, other)) // Shortcut: Equal reference always means equal content
     return true;
   return (object)other != null
     && EqualityContract == other!.EqualityContract
     && EqualityComparer<T>.Default.Equals(FirstProp, other!.FirstProp)
     && ...
}
```

What do you think?

**Comment by HaloFour** [comment]
Honestly, kind of surprised it doesn't do this already.

**Comment by 333fred** [comment]
We had an internal thread based on benchmarks that @PathogenDavid collected here https://github.com/dotnet/csharplang/issues/39#issuecomment-678097433, and had decided to change the implementation. We apparently forgot to follow up on that desire. I'll bring it up with the team and see if we change the behavior in 16.10.

**Reply by alrz** [reply]
I think https://github.com/dotnet/roslyn/issues/47741 is related.

**Reply by 333fred** [reply]
Filed https://github.com/dotnet/roslyn/issues/51136 to follow up on this.
